File: backend/app/services/solver/sudoku_solver.py
import numpy as np


#  Main function
def solve_sudoku(sudoku):
    """
    Solves a Sudoku puzzle using a backtracking algorithm.

    Parameters:
     - sudoku (numpy array): A 9x9 numpy array representing the Sudoku grid,
                            where 0 indicates an empty cell.

    Returns:
     - boolean: True if the Sudoku puzzle is solved, False if it cannot be solved.
    """
    empty_cell = find_next_empty_cell(sudoku)
    if empty_cell is None:
        return True  # Sudoku solved

    row, col = empty_cell

    for value in range(1, 10):
        if is_valid_placement(sudoku, row, col, value):
            sudoku[row, col] = value
            if solve_sudoku(sudoku):
                return True
            sudoku[row, col] = 0  # Backtrack

    return False  # Backtrack if no valid number found


# Supplementary functions 
def find_next_empty_cell(sudoku):
    """
    Finds the next empty cell (with a value of 0) in the Sudoku grid.

    Parameters:
     - sudoku (numpy array): The Sudoku grid.

    Returns:
     - tuple: The row and column index of the next empty cell, or None if no empty cell is found.
    """
    for row in range(9):
        for col in range(9):
            if sudoku[row, col] == 0:
                return row, col
    return None


def is_valid_placement(sudoku, row, col, value):
    """
    Checks if placing a value in a specified cell is valid according to Sudoku rules.

    Parameters:
     - sudoku (numpy array): The Sudoku grid.
     - row (int): Row index of the cell.
     - col (int): Column index of the cell.
     - value (int): The value to check.

    Returns:
     - boolean: True if the placement is valid, False otherwise.
    """
    # Check row and column
    if value in sudoku[row, :] or value in sudoku[:, col]:
        return False

    # Check 3x3 subgrid
    start_row, start_col = 3 * (row // 3), 3 * (col // 3)
    subgrid = sudoku[start_row:start_row + 3, start_col:start_col + 3]
    if value in subgrid:
        return False

    return True

File: sudoku-solver/sudoku-solver/SudokuGridViews/SudokuSolver.swift
//
//  SudokuSolver.swift
//  sudoku-solver
//
//  Created by Molyneux, Piers on 25/01/2024.
//

import Foundation

class SudokuSolver {

    // Main function to solve Sudoku
    func solve(_ sudoku: inout [[Int]]) -> Bool {
        guard let emptyCell = findNextEmptyCell(sudoku) else {
            return true  // Sudoku solved
        }

        let (row, col) = emptyCell

        for value in 1...9 {
            if isValidPlacement(sudoku, row: row, col: col, value: value) {
                sudoku[row][col] = value
                if solve(&sudoku) {
                    return true
                }
                sudoku[row][col] = 0  // Backtrack
            }
        }

        return false  // Backtrack if no valid number found
    }

    // Find the next empty cell in the Sudoku grid
    private func findNextEmptyCell(_ sudoku: [[Int]]) -> (Int, Int)? {
        for row in 0..<9 {
            for col in 0..<9 {
                if sudoku[row][col] == 0 {
                    return (row, col)
                }
            }
        }
        return nil
    }

    // Check if placing a value in a specified cell is valid
    private func isValidPlacement(_ sudoku: [[Int]], row: Int, col: Int, value: Int) -> Bool {
        // Check row and column
        for i in 0..<9 {
            if sudoku[row][i] == value || sudoku[i][col] == value {
                return false
            }
        }

        // Check 3x3 subgrid
        let startRow = 3 * (row / 3)
        let startCol = 3 * (col / 3)
        for i in startRow..<startRow+3 {
            for j in startCol..<startCol+3 {
                if sudoku[i][j] == value {
                    return false
                }
            }
        }

        return true
    }
}
